gh-138223: Fix Infinite loop in email._header_value_parser._fold_mime_parameters when parameter names are too long

ChenyangLi4288 opened the pull request:
The infinite loop occurred in _fold_mime_parameters() when processing MIME parameters
with very long keys (64 characters) during RFC 2231 encoding.

Changes made:
1. In email._header_value_parser._fold_mime_parameters():
   - Replace infinite 'while True:' loop with 'while splitpoint > 1:'
   - Ensure splitpoint is always at least 1 to prevent getting stuck
   - Add fallback logic to force minimal splits when values cannot fit

2. In email.header._append_chunk():
   - Add comment explaining handling of extremely long strings that can't be split

This fixes GitHub issue https://github.com/python/cpython/issues/138223 where add_attachment() with long parameter
keys would cause as_string() to hang indefinitely during MIME parameter
folding and header processing.

<!-- gh-issue-number: gh-138223 -->
* Issue: gh-138223
<!-- /gh-issue-number -->

ChenyangLi4288 committed:
Fix infinite loop in email.message.EmailMessage.as_string() with long parameter keys

The infinite loop occurred in _fold_mime_parameters() when processing MIME parameters
with very long keys (64 characters) during RFC 2231 encoding. The issue was in
two locations:

1. In email._header_value_parser._fold_mime_parameters():
   - Replace infinite 'while True:' loop with 'while splitpoint > 1:'
   - Ensure splitpoint is always at least 1 to prevent getting stuck
   - Add fallback logic to force minimal splits when values cannot fit

2. In email.header._ValueFormatter._append_chunk():
   - Add safety check for extremely long strings that cannot be split
   - Force line breaks when no suitable split points are found
   - Prevent infinite loops in header folding for edge cases

This fixes GitHub issue #138223 where add_attachment() with long parameter
keys would cause as_string() to hang indefinitely during MIME parameter
folding and header processing.

ChenyangLi4288 committed:
Fix infinite loop in email.message.EmailMessage.as_string() with long parameter keys

The infinite loop occurred in _fold_mime_parameters() when processing MIME parameters
with very long keys (64 characters) during RFC 2231 encoding.

Changes made:
1. In email._header_value_parser._fold_mime_parameters():
   - Replace infinite 'while True:' loop with 'while splitpoint > 1:'
   - Ensure splitpoint is always at least 1 to prevent getting stuck
   - Add fallback logic to force minimal splits when values cannot fit

2. In email.header._append_chunk():
   - Add comment explaining handling of extremely long strings that can't be split

This fixes GitHub issue #138223 where add_attachment() with long parameter
keys would cause as_string() to hang indefinitely during MIME parameter
folding and header processing.

picnixz commented on Lib/email/_header_value_parser.py:
```diff
@@ -3081,13 +3081,17 @@ def _fold_mime_parameters(part, lines, maxlen, encoding):
                 # the RFC standard width.
                 maxlen = 78
             splitpoint = maxchars = maxlen - chrome_len - 2
```
maxchars can also be negative here.

ChenyangLi4288 replied:
applied changes to address this and made sure variable scope issue will not happen

picnixz commented on Lib/email/header.py:
```diff
@@ -527,6 +527,8 @@ def _append_chunk(self, fws, string):
                     continue
                 break
             else:
+                # No suitable split point found. Handle the case where we have
```
Remove this comment, it's not necessary.

ChenyangLi4288 replied:
removed

picnixz reviewed (changes requested):
What does the RFC say about folding? Please add a regression test as well and move the blurb entry from library to security.

ChenyangLi4288 committed:
Fix infinite loop in email._header_value_parser._fold_mime_parameters (GH-138223)

- Fix a variable scope issue with encoded_value in _fold_mime_parameters
- Add test case to reproduce and verify the fix
- Update NEWS entry for the security fix

ChenyangLi4288 committed:
trim test script

picnixz commented on Lib/test/test_email/test_infinite_loop_fix.py:
```diff
@@ -0,0 +1,150 @@
+    def test_edge_case_parameter_lengths(self):
+        """Test edge cases around the problematic parameter lengths."""
+        # Test parameter keys of various lengths around the problematic 64-char mark
+        test_lengths = [60, 61, 62, 63, 64, 65, 66, 67, 68]
```
This isn't sufficient. 64 is arbitrary, it depends on the length of the value.

ChenyangLi4288 replied:
I calculated the min_name_len that makes maxchars == 1 now.

picnixz commented on Lib/test/test_email/test_infinite_loop_fix.py:
```diff
@@ -0,0 +1,150 @@
+        """Test that very long parameter keys don't cause infinite loops."""
+        msg = email.message.EmailMessage()
+
+        # Create a parameter key that's 64 characters long (the problematic length)
```
Don't call it a "the problematic length" as it's not absolute.

picnixz commented on Lib/test/test_email/test_infinite_loop_fix.py:
```diff
@@ -0,0 +1,150 @@
+        )
+
+        # Set the long parameter using set_param
+        msg.set_param(long_key, "test_value")
```
Is it really problematic with `test_value`?

picnixz commented on Lib/test/test_email/test_infinite_loop_fix.py:
```diff
@@ -0,0 +1,150 @@
+            # If we get here, no infinite loop occurred
+            self.assertIsInstance(result, str)
+            self.assertIn(long_key, result)
+        except Exception as e:
```
just propagate the errors normally.

picnixz commented on Lib/test/test_email/test_infinite_loop_fix.py:
```diff
@@ -0,0 +1,150 @@
+"""Test for infinite loop fix in email MIME parameter folding.
```
Please, don't add a new test file but extend existing ones.

picnixz reviewed (changes requested):
The tests are incomplete and incorrect as the length limit is dynamic and depends on the length of the parameter's values, and please:

- don't create a new file,
- don't catch unexpected exceptions,
- verify that `main` hangs with the test inputs. The bad values depend on the values.

ChenyangLi4288 committed:
Merge branch 'python:main' into main

ChenyangLi4288 committed:
remove NEWS in Library

ChenyangLi4288 committed:
remove wrong generated file

ChenyangLi4288 committed:
gh-138223: Fix infinite loop in email._header_value_parser._fold_mime_parameters

Fix infinite loop that occurred when processing MIME parameters with very long
parameter names during RFC 2231 encoding. The issue was in the while loop
that tried to find split points for long parameter values.

Changes made:
- In email._header_value_parser._fold_mime_parameters(): Replace infinite
  'while True:' loop with 'while splitpoint > 1:' and ensure splitpoint is
  always at least 1 to prevent getting stuck
- Add fallback logic to force minimal splits when values cannot fit

Testing:
- Added test_mime_parameter_folding_no_infinite_loop to TestMIMEPart class
- Test creates scenario where maxchars = 1 (edge case that caused infinite loop)
- Verifies as_string() completes successfully instead of hanging
- Test passes, confirming the fix prevents infinite loops

This fixes GitHub issue #138223 where add_attachment() with long parameter
keys would cause as_string() to hang indefinitely during MIME parameter
folding and header processing.

ChenyangLi4288 committed:
add NEWS

ChenyangLi4288 committed:
Merge branch 'main' of https://github.com/ChenyangLi4288/cpython
merge from the remote repo

ChenyangLi4288 committed:
Merge branch 'main' into main

picnixz commented on Lib/email/_header_value_parser.py:
```diff
@@ -3080,14 +3080,25 @@ def _fold_mime_parameters(part, lines, maxlen, encoding):
                 if len(encoded_value) <= maxchars:
                     break
                 splitpoint -= 1
+            # If we still can't fit, force a minimal split
```
I think you can use a `while-else` construction here.

picnixz commented on Lib/test/test_email/test_message.py:
```diff
@@ -1087,6 +1087,24 @@ def test_string_payload_with_multipart_content_type(self):
+        result = msg.as_string()
+        self.assertIsInstance(result, str)
+        self.assertIn(long_param_name, result)
 
```
This test is incomplete. Check that you have a line with the param name followed by `=v` with v the first character of the value and check the rest of the lines. IOW, check exactly how the header was formatted.

picnixz commented on Lib/test/test_email/test_message.py:
```diff
@@ -1087,6 +1087,24 @@ def test_string_payload_with_multipart_content_type(self):
+            filename="test.txt"
+        )
+        maxlen = 78 # Standard RFC line length
+        extra_chrome = "utf-8''" # charset + encoding info
```
```suggestion
        maxlen = 78  # Standard RFC line length
        extra_chrome = "utf-8''"  # charset + encoding info
```

picnixz commented on Lib/test/test_email/test_message.py:
```diff
@@ -1087,6 +1087,24 @@ def test_string_payload_with_multipart_content_type(self):
+        section = 0
+        min_name_len = maxlen - 3 - len(str(section)) - 3 - len(extra_chrome)
+        long_param_name = "a" * min_name_len
+        long_param_value = "b" * 100
```
Please test multiple param names and values. Not just min_name_len and 100.

ChenyangLi4288 committed:
Resolve conversation

picnixz commented on Misc/NEWS.d/next/Security/2025-08-29-14-04-17.gh-issue-138223.gX_GJv.rst:
```diff
@@ -0,0 +1,4 @@
+Fix infinite loop in email.message.EmailMessage.as_string() when
+add_attachment() is called with very long parameter keys.  This could cause
+denial of service by hanging the application indefinitely during MIME
+parameter folding and RFC 2231 encoding.
```
```suggestion
:mod:`email.message`: fix an infinite loop :meth:`EmailMessage.as_string
<email.message.EmailMessage.as_string>` when :meth:`~email.message.MIMEPart.add_attachment`
is called with very long parameter keys.

This could cause denial of service by hanging the application indefinitely
during MIME parameter folding and :rfc:`2231` encoding.
```

picnixz commented on Lib/email/_header_value_parser.py:
```diff
@@ -3080,14 +3080,25 @@ def _fold_mime_parameters(part, lines, maxlen, encoding):
                 maxlen = 78
-            splitpoint = maxchars = maxlen - chrome_len - 2
-            while True:
+            maxchars = maxlen - chrome_len - 2
+            # Ensure maxchars is at least 1 to prevent negative values
+            if maxchars <= 0:
+                maxchars = 1
+            splitpoint = maxchars
+            splitpoint = max(1, splitpoint)  # Ensure splitpoint is always at least 1
```
```suggestion
            maxchars = maxlen - chrome_len - 2
            splitpoint = maxchars = max(1, maxchars)
```

picnixz reviewed (commented):
Sorry for not addressing this review! I hope it will help but I think we need an email's maintainer for that (since we force a split...).

bitdancer commented on Lib/email/_header_value_parser.py:
```diff
@@ -3080,14 +3080,25 @@ def _fold_mime_parameters(part, lines, maxlen, encoding):
                 # have that, we'd be stuck, so in that case fall back to
                 # the RFC standard width.
                 maxlen = 78
-            splitpoint = maxchars = maxlen - chrome_len - 2
-            while True:
+            maxchars = maxlen - chrome_len - 2
```
Correct me if I've misunderstood something, but I believe the simpler fix here is the following:

```
-               # have that, we'd be stuck, so in that case fall back to
-               # the RFC standard width.
-               maxlen = 78
+               # the RFC standard width, or the width of the chrome plus
+               # an arbitrary 10 if the parameter name is long.
+               maxlen = max(78, chrome_len + 10
```

With this change applied instead of yours, your test passes, at least.  (I haven't reviewed the test yet; thought I'd post this first).  My logic is that if we are already violating the user's request for maxwidth (and we already have a case where we're doing that), we might as well just make the new value something semi-reasonable, rather than "just enough" (one character more than the chrome).

If this isn't enough to protect against an infinite loop (and I haven't tried to reason through that yet, either), then we are going to need a better test.

ChenyangLi4288 committed:
Apply suggestion from @picnixz

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>
